pkg/controller/constraint/constraint_controller.go

```diff
@@ -193,3 +246,53 @@ func removeString(s string, items []string) []string {
 	}
 	return rval
 }
+
+func NewConstraintsCache() *ConstraintsCache {
+	return &ConstraintsCache{
+		cache: make(map[string]tags),
+	}
+}
+
+func (c *ConstraintsCache) addConstraintKey(constraintKey string, t tags) {
```

Comment:
area for future improvement: we could return a bool for these methods if they actually change the value stored in the map. That bool could be used to toggle reporting metrics. This would avoid unnecessary read locks.

Reply:
sounds good, we can do a follow up